pkg/controllers/machine_controller.go

```diff
@@ -312,92 +312,97 @@ func (r *MachineReconciler) startGarbageCollector(ctx context.Context, log logr.
 
-	err := r.shutdownMachine(ctx, log, machine)
-	if err != nil {
-		if libvirt.IsNotFound(err) {
-			return r.cleanupAdditionalResources(ctx, log, machine)
-		}
-		return fmt.Errorf("failed to shutdown machine: %w", err)
+	if err := r.deleteVolumes(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine disks: %w", err)
 	}
+	log.V(1).Info("Removed machine disks")
 
-	return nil
-}
+	if err := r.deleteNetworkInterfaces(ctx, log, machine); err != nil {
+		return fmt.Errorf("failed to remove machine network interfaces: %w", err)
+	}
+	log.V(1).Info("Removed network interfaces")
 
-func (r *MachineReconciler) destroyDomain(log logr.Logger, machineID string) error {
-	log.V(1).Info("Starting domain destroy")
+	if err := os.RemoveAll(r.host.MachineDir(machine.ID)); err != nil {
+		return fmt.Errorf("failed to remove machine directory: %w", err)
+	}
+	log.V(1).Info("Removed machine directory")
 
-	domain := libvirt.Domain{
-		UUID: libvirtutils.UUIDStringToBytes(machineID),
+	machine.Finalizers = utils.DeleteSliceElement(machine.Finalizers, MachineFinalizer)
+	if _, err := r.machines.Update(ctx, machine); store.IgnoreErrNotFound(err) != nil {
+		return fmt.Errorf("failed to update machine metadata: %w", err)
 	}
+	log.V(1).Info("Removed Finalizers")
 
-	return r.libvirt.DomainDestroy(domain)
+	return nil
 }
 
-func (r *MachineReconciler) shutdownMachine(ctx context.Context, log logr.Logger, machine *api.Machine) error {
-	log.V(1).Info("Shutting down Machine")
+func (r *MachineReconciler) deleteMachine(ctx context.Context, log logr.Logger, machine *api.Machine) (bool, error) {
+	domain := libvirt.Domain{
+		UUID: libvirtutils.UUIDStringToBytes(machine.ID),
+	}
 
 	if machine.Spec.ShutdownAt.IsZero() {
-		machine.Spec.ShutdownAt = time.Now()
-		if _, err := r.machines.Update(ctx, machine); err != nil {
-			return fmt.Errorf("failed to set ShutdownAt: %w", err)
-		}
+		return r.shutdownMachine(ctx, log, machine, domain)
 	}
 
-	domain := libvirt.Domain{
-		UUID: libvirtutils.UUIDStringToBytes(machine.ID),
+	if machine.Spec.ShutdownAt.Add(r.gcVMGracefulShutdownTimeout).Before(time.Now()) {
+		state, _, _, _, _, err := r.libvirt.DomainGetInfo(domain)
+		if libvirt.IsNotFound(err) {
+			return false, nil
+		}
+
+		log.V(1).Info("Wait to complete shutdown", "State", state, "ShutdownAt", machine.Spec.ShutdownAt)
+		return true, nil
 	}
 
-	return r.libvirt.DomainShutdownFlags(domain, libvirt.DomainShutdownAcpiPowerBtn)
+	return false, r.destroyDomain(log, domain)
 }
 
-func (r *MachineReconciler) cleanupAdditionalResources(ctx context.Context, log logr.Logger, machine *api.Machine) error {
-	log.V(1).Info("Starting additional resources cleanup")
-
-	if err := r.deleteVolumes(ctx, log, machine); err != nil {
-		return fmt.Errorf("failed to remove machine disks: %w", err)
+func (r *MachineReconciler) destroyDomain(log logr.Logger, domain libvirt.Domain) error {
+	if err := r.libvirt.DomainDestroy(domain); err != nil {
```

Comment:
pls use graceful Destroy (with flags) benefits is flushing of disk cache before delete machine, so you have small change for FS corruption

Reply:
Will be addressed in the main GC PR